crucible/src/Lang/Crucible/CFG/Extension/Safety.hs

```diff
@@ -0,0 +1,313 @@
+data PartialExpr (ext :: Type)
+                 (e   :: CrucibleType -> Type)
+                 (bt  :: CrucibleType) =
+  PartialExpr
+    { _classifier :: AssertionClassifierTree ext e
+    , _value      :: Maybe (e bt)
+    }
+  deriving (Generic, Typeable)
+
+deriving instance ( Data (e bt)
+                  , Data (e BoolType)
+                  , Data (AssertionClassifier ext e)
+                  , Typeable e
+                  , Typeable ext
+                  , Typeable bt
+                  ) =>
+                  (Data (PartialExpr ext e bt))
+
+deriving instance ( Eq (e bt)
+                  , Eq (e BoolType)
+                  , Eq (AssertionClassifier ext e)
+                  ) =>
+                  (Eq (PartialExpr ext e bt))
+
+deriving instance ( Ord (e bt)
+                  , Ord (e BoolType)
+                  , Ord (AssertionClassifier ext e)
+                  ) =>
+                  (Ord (PartialExpr ext e bt))
+
+deriving instance ( Show (e bt)
+                  , Show (e BoolType)
+                  , Show (AssertionClassifier ext e)
+                  ) =>
+                  (Show (PartialExpr ext e bt))
+
+-- -----------------------------------------------------------------------
+-- *** Lenses
+
+classifier :: Simple Lens (PartialExpr ext e ty) (AssertionClassifierTree ext e)
+classifier = lens _classifier (\s v -> s { _classifier = v})
+
+value :: Simple Lens (PartialExpr ext e ty) (Maybe (e ty))
+value = lens _value (\s v -> s { _value = v })
+
+-- -----------------------------------------------------------------------
+-- *** Instances
+
+instance ( TestEqualityC (AssertionClassifier ext)
+         )
+         => TestEqualityFC (PartialExpr ext) where
+  testEqualityFC :: forall f. (forall x y. f x -> f y -> (Maybe (x :~: y))) ->
+                              (forall x y. PartialExpr ext f x
+                                        -> PartialExpr ext f y
+                                        -> (Maybe (x :~: y)))
+  testEqualityFC subterms (PartialExpr class1 val1) (PartialExpr class2 val2) =
+    let justSubterms x y = isJust (subterms x y)
+    in join (subterms <$> val1 <*> val2) <*
+         guard (liftEq2 justSubterms (testEqualityC subterms) class1 class2)
+
+instance ( OrdC (AssertionClassifier ext)
+         , TestEqualityC (AssertionClassifier ext)
+         )
+         => OrdFC (PartialExpr ext) where
+  compareFC subterms (PartialExpr class1 val1) (PartialExpr class2 val2) =
+    let demote s x y = toOrdering (s x y)
+    in
+      case (val1, val2) of
+        (Just _, Nothing)  -> LTF
+        (Nothing, Just _)  -> GTF
+        (Nothing, Nothing) ->
+          case liftCompare2
+                  (demote subterms) (compareC subterms) class1 class2 of
+            LT -> LTF
+            GT -> GTF
+            EQ -> -- This is safe as long as the 'compareC' is doing a nontrivial
+                  -- comparison, i.e. actually using 'subterms'
+                  unsafeCoerce EQF
```

Comment:
Can we avoid this `unsafeCoerce`? I don't understand the correctness argument you're making here in this comment, and it seems like it depends on the `OrdC` instance in a nontrivial way that makes me uneasy.

Reply:
Yeah, the issue is that we don't have the type-level evidence we need in this branch, but we can't say these things are unequal. This is a pretty fundamental issue because of the shift from `PartExr p v ~= Maybe (p, v)` to a type more like `(p, Maybe v)`. I wonder if we can get rid of the `Maybe` altogether, there should be reasonable "zero" values for each type... I'll look into that.

Reply:
I've done away with the `Maybe` here, see "What4.Partial". This actually led me to combine the ideas from a few different types expressing partiality, and I think leads to a better overall solution.